### PR TITLE
Python: Automated subclass models

### PR DESCRIPTION
**Pull request by RasmusWL**
### Overview

This PR adds automatically captured subclass information for a the majority of interesting PyPI packages. As an example of what this PR achieves:

* the [PyPI package `flask-restplus`](https://pypi.org/project/flask-restplus/) defines `flask_restful.Resource` ([src](https://github.com/noirbizarre/flask-restplus/blob/1fe65ddae4c04315bd88535e3c2eaff381436270/flask_restplus/resource.py#L13)) which is a subclass of `flask.views.MethodView`.
* Our modeling of flask based remote-flow-sources [ultimately depends on being able to figure out whether a class is a subclass of `flask.views.MethodView`](https://github.com/github/codeql/blob/1bc6f88f58e751c5f0dbda151b5959d76c6a150f/python/ql/lib/semmle/python/frameworks/Flask.qll#L243-L257)
* so knowing a subclass of `flask_restful.Resource` is also a `flask.views.MethodView` allows us to model the remote-flow-sources properly.

We've traditionally had automatic dependency installation available on the codeql-action, but are moving to a solution without this (see https://github.blog/changelog/2023-07-12-code-scanning-with-codeql-no-longer-installs-python-dependencies-automatically-for-new-users/).

We've previously relied on analyzing installed dependencies to reach this conclusion (by following subclass relationship). This PR is the solution to still be able to reach the same conclusion when we [stop installing dependencies](https://github.blog/changelog/2023-07-12-code-scanning-with-codeql-no-longer-installs-python-dependencies-automatically-for-new-users/).

We achieve this by using the extensible type-models to ahead-of-time record important subclass/aliasing information. For example, see the very first commit (2f17d2f)

Our internal testing shows that for all but a few cases, we end up with a solution comparable or better to what we had before, even when narrowing our focus to repos where dependency installation was successful before.

(Thanks to @tausbn for helping with the modeling :heart:)

### Reviewing this PR

What a mess. Sorry. It's a mix of working on the tooling `process-mrva-results.py`/`SubclassFinder.qll` and enabling subclasses/aliases to be found in the actual modeling. The latter bit required removing the `private` annotations for much of our modeling. I think we'll just have to live with that.

The only commits I've found that don't follow this pattern, and that could have been made into separate PRs, are:

* ba19f95
* 5920a8f

### Notes

1. The tooling to generate these subclass-capture models automatically only lives internally.
2. to ensure this automated modeling could still be recreated once we don't do dependency installation (if we wanted to use a different format say), the actual modeling with MRVA has only been done while I made sure we wouldn't make use of any dependencies that might have been installed (specifically by [this commit](https://github.com/github/codeql/commit/83f6460fc03612fe734755086200d9f2440b502f)).

### CODE REVIEW

**Review by tausbn — DISMISSED** [review]
Only a few suggestions here and there, otherwise I think this looks good!

Of course, we should do some performance testing before merging. Also, it'll be interesting to see if all of these models make a difference in terms of results for our standard suite.

File: python/ql/src/meta/ClassHierarchy/join-yml-files.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+as_lists.sort()
+
+
+to_write = wrap_in_template(as_lists)
+write_data(to_write, joined_file)
+
+print("Joined all files into", joined_file)
+
+for f in glob.glob(f"{subclass_capture_path}/auto-*.model.yml", recursive=True):
+    os.unlink(f)
```
Comment (yoff) [inline python/ql/src/meta/ClassHierarchy/join-yml-files.py:30]: I wonder if users might be surprised by this also being part of the script behaviour? It should probably be mentioned in the doc-string at least.

File: python/ql/src/meta/ClassHierarchy/split-yml-files.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+if not joined_file.exists():
+    sys.exit(f"File {joined_file} does not exists")
+
+all_data = parse_from_file(joined_file)
+package_data = defaultdict(set)
+for t in all_data:
+    package_data[t[1]].add(t)
+write_all_package_data_to_files(package_data)
+
+joined_file.unlink()
```
Comment (yoff) [inline python/ql/src/meta/ClassHierarchy/split-yml-files.py:22]: Similarly to above. The modality of "either you have the joined file or you have all the split ones" should probably be made clear.

**Review by yoff — DISMISSED** [review]
I would suggest a few comments, but I understand that these may be obvious for people working with the tooling, so I am willing to approve this as is.

File: python/ql/src/meta/ClassHierarchy/join-yml-files.py (hunk 1 above)
Comment (RasmusWL) [reply python/ql/src/meta/ClassHierarchy/join-yml-files.py:30]: I've clarified in a local commit, but do note that "users" are only internal codeql-developers right now.

**Comment by RasmusWL** [comment]
I've had to fix the query to find new subclasses, since we did not account for `rest_framework.response.Response` being a subclass of `django.http.response.HttpResponse` in the spec definition. This meant that in the .yml files for any subclass of the rest_framework Response, it would be modeled as BOTH rest_framework and django responses. 

However, since the rest_framework modeling of responses is a more specific subclass of django responses (see code below), we actually only want our .yml modeling to contain the row for `rest_framework` responses.

https://github.com/github/codeql/blob/43fe9ca31d4421bfe18b2bf6116e8abb285f062a/python/ql/lib/semmle/python/frameworks/RestFramework.qll#L312-L325

I also realized that we had missed a few important repos in our automatic analysis.

After fixing those two problems, I have regenerated all the modeling, and force pushed some updates to this branch rewriting history to remove the old modeling commits. I guess I am worried about adding multiple commits that each changes 80k lines, so everything has been squashed together instead. I can refer to individual SHAs or restore the actual history if a reviewer would find that helpful :blush:

The `git diff --shortstat` for the old vs. new modeling is `419 files changed, 12226 insertions(+), 12955 deletions(-)`. I did some looking through the diff manually, and it seems like quite a few is the more precise modeling of django rest framework response (so it's not ALSO modeled as a pure Django Response in the .yml files). However, there is also quite a few other changes, which seems to be due to what repos actually had DBs available for MRVA analysis. This suggest it would be smart to keep track of exactly what repos were analyzed at what SHAs, so it's easier to debug such situations in the future.

Internal note: The results were achieved after 5 rounds of analysis of the list of repos (which is internal only). In the last round, only 1 additional model was added, so it should be fine to stick with 4 rounds next time :blush:

**Review by tausbn — APPROVED** [review]
Overall I think this looks good to me. There's a few places where we still use `classRef` or `cls` instead of `subclassRef`, but fixing this inconsistency is probably best left for a follow-up PR.

**Pull request merged**